options/posix/generic/grp-stubs.cpp

```diff
@@ -127,22 +173,65 @@ struct group *getgrent(void) {
 	__builtin_unreachable();
 }
 struct group *getgrgid(gid_t gid) {
-	return walk_file([&] (group *entry) {
+	int err = walk_file(&global_entry, [&] (group *entry) {
 		return entry->gr_gid == gid;
 	});
+
+	if (err) {
+		errno = err;
+		return nullptr;
+	} else {
+		return &global_entry;
+	}
 }
-int getgrgid_r(gid_t, struct group *, char *, size_t, struct group **) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+int getgrgid_r(gid_t gid, struct group *grp, char *buffer, size_t size, struct group **result) {
+	*result = nullptr;
+	int err = walk_file(grp, [&] (group *entry) {
+		return entry->gr_gid == gid;
+	});
+
+	if (err) {
+		return err;
+	}
+
+	err = copy_to_buffer(grp, buffer, size);
+	if (err) {
+		return err;
+	}
+
+	*result = grp;		
+	return 0;
 }
 struct group *getgrnam(const char *name) {
-	return walk_file([&] (group *entry) {
+	int err = walk_file(&global_entry, [&] (group *entry) {
 		return !strcmp(entry->gr_name, name);
 	});
+
+	if (err) {
+		errno = err;
+		return nullptr;
+	} else {
```

Comment:
LGTM although minor style point is the unecessary else-case here and once more above.